### PR TITLE
implement token transfer feature

### PR DESCRIPTION
**Pull request by LogvinovLeon**
This PR:
* Adds the posibility to transfer tokens directly from 0x Portal
screen shot 2017-09-07 at 12 05 37
screen shot 2017-09-07 at 12 06 01
screen shot 2017-09-07 at 12 06 17
screen shot 2017-09-07 at 12 09 33
screen shot 2017-09-07 at 12 09 43
screen shot 2017-09-07 at 12 09 56
screen shot 2017-09-07 at 12 10 59

### CODE REVIEW

**Comment by fabioberger** [comment]
Let's not await for it to be mined. Let's create a link to EtherScan.io with txHash and show it in flashMessage. Also, make flashMessage stick around for 15sec.

**Comment by LogvinovLeon** [comment]
I'd rather do both. We still need to update the balances after it's mined.
We can show the flashMessage immidiate then.

File: ts/blockchain.ts [hunk 1]
```diff
@@ -181,6 +181,13 @@ export class Blockchain {
         const allowance = amountInBaseUnits;
         this.dispatcher.replaceTokenAllowanceByAddress(token.address, allowance);
     }
+    public async transferAsync(tokenAddress: string, toAddress: string,
+                               amountInBaseUnits: BigNumber.BigNumber): Promise<void> {
+        const txHash = await this.zeroEx.token.transferAsync(
+            tokenAddress, this.userAddress, toAddress, amountInBaseUnits,
+        );
+        await this.zeroEx.awaitTransactionMinedAsync(txHash);
```
Comment (fabioberger) [inline ts/blockchain.ts]: Let's not await, but show Etherscan link instead!

File: ts/components/token_balances.tsx [hunk 2]
```diff
@@ -383,6 +387,16 @@ export class TokenBalances extends React.Component<TokenBalancesProps, TokenBala
...
+                <TableRowColumn
+                    style={{paddingLeft: actionPaddingX, paddingRight: actionPaddingX}}
+                >
+                    <TransferButton
+                        blockchain={this.props.blockchain}
+                        dispatcher={this.props.dispatcher}
+                        token={token}
+                        onError={this.onEthWethConversionFailed.bind(this)}
+                    />
+                </TableRowColumn>
```
Comment (fabioberger) [inline ts/components/token_balances.tsx]: How does this look on mobile?

File: ts/components/token_balances.tsx [hunk 3]
```diff
@@ -519,6 +533,30 @@ export class TokenBalances extends React.Component<TokenBalancesProps, TokenBala
             return false;
         }
     }
+    private async onTransfer(token: Token): Promise<boolean> {
+        try {
+            await this.props.blockchain.mintTestTokensAsync(token);
```
Comment (fabioberger) [inline ts/components/token_balances.tsx]: Why are we minting tokens `onTransfer`???

File: ts/components/transfer_button.tsx [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+interface TransferButtonProps {
+    token: Token;
+    dispatcher: Dispatcher;
+    blockchain: Blockchain;
+    onError: () => void;
+}
+
+interface TransferButtonState {
+    isTransferDialogVisible: boolean;
+    isTransferHappening: boolean;
```
Comment (fabioberger) [inline ts/components/transfer_button.tsx]: let's rename this to `isTransferring`.

File: ts/components/transfer_button.tsx [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    }
+    public render() {
+        const labelStyle = this.state.isTransferHappening ? {fontSize: 10} : {};
+        return (
+            <div>
+                <RaisedButton
+                    style={{width: '100%'}}
+                    labelStyle={labelStyle}
+                    disabled={this.state.isTransferHappening}
+                    label={this.state.isTransferHappening ? 'Sending...' : 'Send'}
```
Comment (fabioberger) [inline ts/components/transfer_button.tsx]: Why are we not using the `LifecycleRaisedButton` here?

File: ts/components/transfer_button.tsx [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    private async onTransferAmountSelectedAsync(recipient: string, value: BigNumber.BigNumber) {
+        this.setState({
+            isTransferHappening: true,
+        });
+        this.toggleTransferDialog();
+        const token = this.props.token;
+        let balance = token.balance;
+        try {
+            await this.props.blockchain.transferAsync(token.address, recipient, value);
+            // TODO
```
Comment (fabioberger) [inline ts/components/transfer_button.tsx]: TODO what?

File: ts/components/transfer_button.tsx [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+            const tokenAmount = ZeroEx.toUnitAmount(value, token.decimals);
+            const flashMessage = `Successfully transfered ${tokenAmount.toString()} ${token.symbol} to ${recipient}`;
```
Comment (fabioberger) [inline ts/components/transfer_button.tsx]: typo: `transfered` should be `transferred`

File: ts/components/transfer_button.tsx [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+            this.props.dispatcher.showFlashMessage(flashMessage);
+            balance = balance.minus(value);
+            const updatedToken = _.assign({}, token, {
+                balance,
+            });
+            this.props.dispatcher.updateTokenByAddress([updatedToken]);
+        } catch (err) {
+            const errMsg = '' + err;
```
Comment (fabioberger) [inline ts/components/transfer_button.tsx]: Let's use template string here please.

File: ts/components/token_balances.tsx (hunk 2 above)
Comment (LogvinovLeon) [reply ts/components/token_balances.tsx]: screen shot 2017-09-07 at 16 54 24
Could be worse ;)

File: ts/components/transfer_button.tsx (hunk 5 above)
Comment (LogvinovLeon) [reply ts/components/transfer_button.tsx]: For the same reason we don't do that in Ether -> Weth conversion.
Here the action is not just an async function.
It's opening a dialog, selecting smth in a dialog and submitting the transaction.
It's much easier to do it with state.

**Comment by fabioberger** [comment]
Bug: text should be center aligned:

screen shot 2017-09-07 at 19 29 28

**Comment by fabioberger** [comment]
Bug: Mobile button overlap... maybe remove on mobile?

screen shot 2017-09-07 at 19 30 24

**Comment by fabioberger** [comment]
Suggestion: Let's still have the link to Etherscan in success flash message. People might still want to verify the tx when through on a third-party.

**Comment by fabioberger** [comment]
Bug: closing and re-opening the dialog has the sender address persist:

screen shot 2017-09-07 at 19 31 47

After a successful send, the address field should be cleared.

**Comment by fabioberger** [comment]
Bug: The dialog changes height when inputting an address:

transfer_dialog

File: ts/components/transfer_dialog.tsx [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+            <FlatButton
+                disabled={this.hasErrors()}
+                label="Transfer"
+                primary={true}
+                onTouchTap={this.onTransferClick.bind(this)}
+            />,
+        ];
+        return (
+            <Dialog
+                title="I want to tranfer"
```
Comment (fabioberger) [inline ts/components/transfer_dialog.tsx]: Let's standardize around using the word `Send` everywhere.

**Comment by fabioberger** [comment]
Nit: Flash message not always centered.
Nit two -- not enough padding to right of button:

screen shot 2017-09-07 at 19 42 29

File: ts/components/transfer_button.tsx [hunk 10]
```diff
@@ -0,0 +1,92 @@
+import * as _ from 'lodash';
+import {ZeroEx} from '0x.js';
+import * as React from 'react';
+import * as BigNumber from 'bignumber.js';
+import RaisedButton from 'material-ui/RaisedButton';
+import {BlockchainCallErrs} from 'ts/types';
+import {TransferDialog} from 'ts/components/transfer_dialog';
+import {Side, Token} from 'ts/types';
```
Comment (fabioberger) [inline ts/components/transfer_button.tsx]: `Side` is unused.

File: ts/components/transfer_button.tsx [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+            this.props.dispatcher.showFlashMessage(flashMessage);
+            balance = balance.minus(value);
+            const updatedToken = _.assign({}, token, {
+                balance,
+            });
+            this.props.dispatcher.updateTokenByAddress([updatedToken]);
+        } catch (err) {
+            const errMsg = `${err}`;
+            if (_.includes(errMsg, BlockchainCallErrs.USER_HAS_NO_ASSOCIATED_ADDRESSES)) {
+                this.props.dispatcher.updateShouldBlockchainErrDialogBeOpen(true);
```
Comment (fabioberger) [inline ts/components/transfer_button.tsx]: Let's put a return after this just to make this a bit more future-proof.

File: ts/components/transfer_button.tsx [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+            } else if (!_.includes(errMsg, 'User denied transaction')) {
+                utils.consoleLog(`Unexpected error encountered: ${err}`);
+                utils.consoleLog(err.stack);
+                await errorReporter.reportAsync(err);
+                this.props.onError();
```
Comment (fabioberger) [inline ts/components/transfer_button.tsx]: And a return here too.

File: ts/utils/utils.ts [hunk 13]
```diff
@@ -134,13 +134,13 @@ export const utils = {
         const isUserOnMobile = isMobile();
         return isUserOnMobile;
     },
-    getEtherScanLinkIfExists(address: string, networkId: number, suffix: EtherscanLinkSuffixes): string {
+    getEtherScanLinkIfExists(addressOrTxHash: string, networkId: number, suffix: EtherscanLinkSuffixes): string {
```
Comment (fabioberger) [inline ts/utils/utils.ts]: Nice fix!

File: ts/components/transfer_button.tsx (hunk 10 above)
Comment (LogvinovLeon) [reply ts/components/transfer_button.tsx]: Did you install WebStorm? XD

File: ts/components/flash_messages/token_transfer_completed.tsx [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+            null :
+            (
+                <a
+                    style={{color: 'white'}}
+                    href={`${this.props.etherScanLinkIfExists}`}
+                    target="_blank"
+                >
+                    Verify on Etherscan
+                </a>
+            );
```
Comment (fabioberger) [inline ts/components/flash_messages/token_transfer_completed.tsx]: Why not put this in the jsx like we always do:

```
{!_.isUndefined(this.props.etherScanLinkIfExists) &&
      <a
         style={{color: 'white'}}
         href={`${this.props.etherScanLinkIfExists}`}
         target="_blank"
        >
            Verify on Etherscan
        </a>
}
```

File: ts/components/flash_messages/token_transfer_completed.tsx [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+import {utils} from 'ts/utils/utils';
+
+interface TokenTransferCompletedProps {
+    etherScanLinkIfExists?: string;
+    token: Token;
+    toAddress: string;
+    amountInBaseUnits: BigNumber.BigNumber;
+}
+
+interface TokenTransferCompletedState {}
```
Comment (fabioberger) [inline ts/components/flash_messages/token_transfer_completed.tsx]: Let's rename to `TokenSendCompleted`

File: ts/blockchain.ts [hunk 16]
```diff
@@ -177,10 +182,24 @@ export class Blockchain {
...
         this.dispatcher.replaceTokenAllowanceByAddress(token.address, allowance);
     }
+    public async transferAsync(token: Token, toAddress: string,
+                               amountInBaseUnits: BigNumber.BigNumber): Promise<void> {
+        const txHash = await this.zeroEx.token.transferAsync(
+            token.address, this.userAddress, toAddress, amountInBaseUnits,
+        );
+        await this.showEtherScanLinkAndAwaitTransactionMinedAsync(txHash);
+        const etherScanLinkIfExists = utils.getEtherScanLinkIfExists(txHash, this.networkId, EtherscanLinkSuffixes.tx);
+        this.dispatcher.showFlashMessage(React.createElement(TokenTransferCompleted, {
```
Comment (fabioberger) [inline ts/blockchain.ts]: Why do we use `React.createElement` here? We always just use JSX:

```
this.dispatcher.showFlashMessage((
  <TokenTransferCompleted
    etherScanLinkIfExists={etherScanLinkIfExists}
    ....
  />
))
```

Comment (fabioberger) [reply ts/blockchain.ts]: I guess your way is more succinct... it's fine.

File: ts/utils/utils.ts [hunk 17]
```diff
@@ -195,4 +195,8 @@ export const utils = {
         const id = name.replace(/ /g, '-');
         return id;
     },
+    truncateAddress(address: string): string {
```
Comment (fabioberger) [inline ts/utils/utils.ts]: Let's rename this to `getAddressBeginAndEnd`

File: ts/blockchain.ts (hunk 16 above)
Comment (LogvinovLeon) [reply ts/blockchain.ts]: The real reason - is that this file is not jsx

**Pull request merged**